[fix] Fixes missing probabilities and unreachable rules in PCFG CNF conversion

ruoxining opened the pull request:
This PR is about several fixes relating to the CNF conversion of PCFGs wrt. Issue #3384 .

Changes:
- Add symbols `@` and `$` in the default padding symbol `@$@` into the recognizing regex (Line 1701).
- The previous CNF conversion in PCFG will have all the possibilities lost. This fix maintains the probabilities after the CNF conversion.
- Unreachable rules were removed after the unitary conversion. For example, if a grammar goes as `S -> 'a' [1.0] | A -> 'a' [1.0]` and `S` is the start symbol, we only maintain the grammar as `S -> 'a' [1.0]` since A is never reached. Otherwise, these redundant rules will cause extra complexity when parsing sentences with this grammar.
- Unreachable rules are also removed in the CFG class.
- `nltk/test/grammar.doctest` is modified according to the changes above.

Notes:
- pre-commit test passed.
- related pytests` grammar.doctest` and `grammartestsuites.doctest` passed.

ruoxining committed:
[fix] fix regex to match the new default symbols

ruoxining committed:
[fix] fix pcfg unary reduction

ruoxining committed:
[doc] add docstrings

ruoxining committed:
Merge branch 'nltk:develop' into develop

ruoxining committed:
[feat] add removing mixed rules for pcfg

ruoxining committed:
[feat] fix unreachable rules for cfg

ruoxining committed:
[fix] fix space related style

ruoxining committed:
[chore] add author list

ruoxining committed:
[fix] fix regular expression

ruoxining committed:
[fix] fix number of rules in unit test

ruoxining committed:
[feat] add related unit tests

ruoxining commented on nltk/test/grammar.doctest:
```diff
@@ -85,4 +85,43 @@ Grammars with mixed rules can be converted into Chomsky Normal Form:
     >>> grammar
     <Grammar with 15 productions>
     >>> grammar.chomsky_normal_form()
-    <Grammar with 18 productions>
```
`Det -> 'a' | 'the'` has no preceding generation rule and are thus removed.

ruoxining commented on nltk/test/grammar.doctest:
```diff
@@ -85,4 +85,43 @@ Grammars with mixed rules can be converted into Chomsky Normal Form:
+    >>> grammar
+    <Grammar with 5 productions>
+    >>> grammar.chomsky_normal_form()
+    <Grammar with 4 productions>
```
`[V -> 'v' [1.0], S -> N S@$@N [1.0], S@$@N -> V P [1.0], P -> 'p' [1.0]]`

ruoxining commented on nltk/test/grammar.doctest:
```diff
@@ -85,4 +85,43 @@ Grammars with mixed rules can be converted into Chomsky Normal Form:
+    ... A -> 'a' [1.0]
+    ... """)
+    >>> grammar
+    <Grammar with 3 productions>
```
`[S@$@A -> S A [1.0], S -> 's' [1.0], A -> 'a' [1.0]]`

stevenbird commented:
@ruoxining would you please open an issue to document the problems that are resolved by this PR?

ruoxining commented:
> @ruoxining would you please open an issue to document the problems that are resolved by this PR?

Hi sure! I've created one here #3384

stevenbird commented:
@ruoxining can we also please have some tests to demonstrate handling of non-trivial probability manipulations, e.g.:

```
S -> NP VP PP [0.5]
S -> NP VP [0.5]
etc
```

Which could produce something like:

```
S -> NP X [1.0]
X -> VP Y [1.0]
Y ->
etc
```

Another test could ensure the productions from any non-terminal have probability summing to 1.

Thanks

ruoxining commented:
> @ruoxining can we also please have some tests to demonstrate handling of non-trivial probability manipulations, e.g.:
> 
> ```
> S -> NP VP PP [0.5]
> S -> NP VP [0.5]
> etc
> ```
> 
> Which could produce something like:
> 
> ```
> S -> NP X [1.0]
> X -> VP Y [1.0]
> Y ->
> etc
> ```
> 
> Another test could ensure the productions from any non-terminal have probability summing to 1.
> 
> Thanks

Sure @stevenbird! Thanks for your test - it caught a problem in my probability definition. I'm fixing it and pushing the code along with more tests soon.

ruoxining committed:
[doc] fix examples

ruoxining committed:
[fix] normalize probabilities after removing

ruoxining committed:
[feat] add tests for non-trivial probability

ruoxining committed:
[doc] remove todo

ruoxining committed:
[fix] merge repeated grammars

ruoxining committed:
[feat] update test case requiring merging

ruoxining committed:
[fix] modify test case to require removing unreachable rule

ruoxining commented:
> @ruoxining can we also please have some tests to demonstrate handling of non-trivial probability manipulations, e.g.:
> 
> ```
> S -> NP VP PP [0.5]
> S -> NP VP [0.5]
> etc
> ```
> 
> Which could produce something like:
> 
> ```
> S -> NP X [1.0]
> X -> VP Y [1.0]
> Y ->
> etc
> ```
> 
> Another test could ensure the productions from any non-terminal have probability summing to 1.
> 
> Thanks

Hi @stevenbird I've modified the tests and added a check for probabilities summing to 1!

The definitions of probabilities for each conversion rule:
- START: 
    Add a new start symbol `S0` pointing to the previous start symbol `S0 -> S [1.0]`.

- TERM:
    For a mixed rule `S -> 'a' B C [p]`, add a new symbol `A` and turn the rules into `S -> A B C [p], A -> 'a' [1.0]`.

- MUL:
    For a rule with multiple RHSes `S -> A B C D [p]`, recursively add new symbols and turn it into `S -> A S@$@A [p], S@$@A -> B S@$@A@$@B [1.0], S@$@A@$@B -> C D [1.0]`.

- UNIT: (the most complicated one affecting almost all downstream probabilities)
    For a grammar with a unitary rule
    ```
     A -> B [0.3] | B C [0.7] 
     B -> C [0.8] | B C [0.2]
     C -> 'c' [0.5] | D [0.5]
     D -> 'd' [1.0]
    ```
    (1) Replace unitary rules by multiplying the probabilities of this rule and the downstream rules
    ```
     A -> 'c' [0.3 * 0.8 * 0.5 = 0.12] | 'd' [0.3 * 0.8 * 0.5 = 0.12] | B C [0.3 * 0.2 = 0.06] | B C [0.7]
     B -> 'c' [0.8 * 0.5 = 0.4] | 'd' [[0.8 * 0.5 = 0.4] | B C [0.2]
     C -> 'c' [0.5] | 'd' [0.5 * 1.0 = 0.5]
     D -> 'd' [1.0]
    ```
    (2) Merge the same rules
    ```
     A -> 'c' [0.12] | 'd' [0.12] | B C [0.7 + 0.06 = 0.76]
     B -> 'c' [0.4] | 'd' [0.4] | B C [0.2]
     C -> 'c' [0.5] | 'd' [0.5]
     D -> 'd' [1.0]
    ```
    (3) Remove unreachable rules.
    ```
     A -> 'c' [0.12] | 'd' [0.12] | B C [0.76]
     B -> 'c' [0.4] | 'd' [0.4] | B C [0.2]
     C -> 'c' [0.5] | 'd' [0.5]
    ```

The probability changes are also demonstrated in the function docstrings.

I actually didn't find sufficient material on handling the probabilities in PCFG's CNF conversion, so I mostly 'come up with' these procedures - but cases I didn't think of also still have a chance to exist. Hope everything is correct, and double-checkings are welcomed!

ruoxining commented:
Hi @stevenbird I saw you self-assigned for this pull request, but if you have any concern or question, please feel free to tell me to change or discuss. Thanks for your efforts!

stevenbird commented:
@ruoxining thanks for all this work, and careful documentation of cases. Yes, in the absence of a published algorithm, you have had to do original work here. I expect there are open source implementations for converting PCFGs to CNF, so I wonder if you could please see if you can find any, just so we can be confident you have covered all cases?

stevenbird commented:
@ruoxining, it's good to have this contribution, however I'm sure you're not the first person to come up with an algorithm for this task. So it would be good if you could please check if an algorithm already exists.

ruoxining commented:
@stevenbird Hi Steven, thanks for this information. I did find related papers and slides https://arxiv.org/pdf/1201.2374, and they took me a while to read. I will proceed to check and modify my implementation in recent several days.

stevenbird commented:
@ruoxining, would you please open a new PR when you're ready?
This will be a welcome contribution.
